### PR TITLE
text cursor hover

### LINKED ISSUE #46: Change cursor when  hovering text widget
I beam cursor when hovered, revert when not.

### PR DESCRIPTION
**Pull request by bytemunch**
closes #46 

An example of how hover-detection and cursor changing can happen. Interaction functions are in userland, should hover join them there or be inbuilt?

PR currently expands interaction fns in userland to add hover bits

### CODE REVIEW

**Comment by Dimchikkk** [comment]
What do you think about having plugin config property to set cursors for `onTyping`, `onHover` and when cursor out of the edit box with the defaults: onTyping: Cursor::None, onHover: Cursor::Text, onOutOfBox: Cursor::Default?

Maybe with option to disable cursor handling fully.

Edit: maybe userland is fine :)

**Comment by bytemunch** [comment]
I think there's arguments for and against. For UX it's better to have clear feedback that shows the box is editable, but a lot of games have custom cursors and will want to handle the events on their end. I feel having hover cursors is a sane default, and a config option to provide custom cursors or ignore all cursor modification is a good idea for other use cases.

Not sure about `onTyping` hiding the cursor, do you have a use case for that?

**Comment by Dimchikkk** [comment]
@bytemunch type on github text input...giving the comment... it hides the mouse cursor on typing :)
(the same for google translator input)

**Comment by bytemunch** [comment]
Ah ok, doesn't happen on my Linux, will try with Windows tomorrow and see if I can find out what OSX does. If it's normal UX then emulating it should be default, I agree

**Comment by Dimchikkk** [comment]
Oh, I am on Mac OS, the mouse cursor is hidden on typing _this_ comment.

**Comment by bytemunch** [comment]
Okay, so definitely hide on OSX, Windows is (expectedly) a [mess](https://superuser.com/questions/928839/what-does-the-hide-pointer-while-typing-feature-actually-do) and Linux is... Linux. I think hide on typing by default makes aesthetic sense, can't go wrong following Apple's lead

**Comment by bytemunch** [comment]
added hover events for extensibility with custom cursors, and a config option to disable inbuilt cursor modification

File: src/lib.rs [hunk 1]
```diff
@@ -407,7 +419,17 @@ impl Plugin for CosmicEditPlugin {
                 swash_cache: SwashCache::new(),
             })
             .insert_resource(CosmicFontSystem(font_system))
+            // Cursor Bits
```
Comment (Dimchikkk) [inline src/lib.rs]:
```suggestion
       
```

File: src/lib.rs [hunk 2]
```diff
@@ -407,7 +419,17 @@ impl Plugin for CosmicEditPlugin {
...
             .add_event::<CosmicTextChanged>();
+
+        // Cursor Bits
+        app.add_systems(Update, (hover_sprites, hover_ui))
```
Comment (Dimchikkk) [inline src/lib.rs]: Let's hide it under the `change_cursor` flag too.

File: src/cursor.rs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            Interaction::None => {
+                evw_hover_out.send(TextHoverOut);
+            }
+            Interaction::Hovered => {
+                evw_hover_in.send(TextHoverIn);
+            }
+            _ => {}
+        }
+        if let Interaction::Pressed = interaction {
+            commands.insert_resource(Focus(Some(entity)));
```
Comment (Dimchikkk) [inline src/cursor.rs]:
```suggestion
```
Let's not handle Focus on plugin side.

File: src/lib.rs (hunk 2 above)
Comment (bytemunch) [reply src/lib.rs]: They emit the events for on and off hover, so if user wants to make a custom cursor implementation those fns will still be needed. Maybe more granularity on the options, `change_cursor: default | events | none`?

File: src/cursor.rs (hunk 3 above)
Comment (bytemunch) [reply src/cursor.rs]: my bad, copy pasted the fn and missed that bit >.<

File: src/lib.rs (hunk 2 above)
Comment (Dimchikkk) [reply src/lib.rs]: For custom cursors I would prefer plugin config options for cursors.... but `change_cursor` with `default | events | none` also should be fine 👍 I just want to have an option to fully disable it if needed.

File: src/cursor.rs (hunk 3 above)
Comment (Dimchikkk) [reply src/cursor.rs]: no prob :)

**Review by Dimchikkk — APPROVED** [review]
nice!

File: src/lib.rs (hunk 2 above)
Comment (bytemunch) [reply src/lib.rs]: Custom cursor configs would be simpler if bevy_cosmic_edit was the only one controlling the cursor, but I feel if there are other plugins or code that also need cursor control hooking an event would be easier to integrate. Full disable a good shout, no need for unused events

Comment (Dimchikkk) [reply src/lib.rs]: Makes sense

**Pull request merged**